multi-network policy support

cathy-zhou opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

**- What this PR does and why is it needed**
<!--
A summary of the changes within this pull request and some context
as to why they were made
-->

**- Special notes for reviewers**
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->


**- How to verify it**
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

cathy-zhou commented:
/retest

cathy-zhou commented:
/retest

maiqueb reviewed (commented):
Some nits I've found until now (I couldn't finish reviewing the entire PR).

Would it make sense to split the last commit in two ? One adding the "feature", and another adding the ACL logging ? I kept getting distracted by that while reviewing.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_multipolicy.go:
```diff
@@ -0,0 +1,123 @@
+const PolicyForAnnotation = "k8s.v1.cni.cncf.io/policy-for"
+
+func (bnc *BaseNetworkController) syncMultiNetworkPolicies(multiPolicies []interface{}) error {
+	expectedPolicies := make(map[string]map[string]bool)
```
this is the same as latest syncNetworkPolicies today

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,267 @@
+// This function implements the main body of work of syncNamespaces.
+// Upon failure, it may be invoked multiple times in order to avoid a pod restart.
+func (bnc *BaseNetworkController) syncNamespaces(namespaces []interface{}) error {
+	expectedNs := make(map[string]bool)
```
this is old code, I didn't change this.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_multipolicy.go:
```diff
@@ -0,0 +1,123 @@
+		a := strings.Split(networkName, "/")
+		if len(a) > 1 {
+			networkName = a[1]
+			networkNamespace = a[0]
```
according to the existing implementation based on iptable, that is how it works:of the https://github.com/k8snetworkplumbingwg/multi-networkpolicy-iptables/blob/master/pkg/server/server.go

cathy-zhou commented on go-controller/pkg/libovsdbops/address_set.go:
this is used by cleanupPolicyLogicalEntities()

jcaamano commented on go-controller/pkg/network-controller-manager/network_controller_manager.go:
```diff
@@ -242,10 +244,49 @@ func (cm *networkControllerManager) configureMetrics(stopChan <-chan struct{}) {
+	return nil
+}
+
+func (cm *networkControllerManager) configureAclLoggingSupport() {
```
Yeah, why not handle the error as we do for example `enableOVNLogicalDataPathGroups`?
Multicast is different because what we are doing there is checking if OVN supports it, but here it is just a random error?

cathy-zhou replied:
@jcaamano @maiqueb  I am keeping the current behavior for default network. see https://github.com/ovn-org/ovn-kubernetes/blob/43da7b1a7c62ba78fffa14965a4bee552d24dc11/go-controller/pkg/ovn/default_network_controller.go#L326

jcaamano reviewed (commented):
just reviewed the first few files. 
@cathy-zhou is https://github.com/ovn-org/ovn-kubernetes/pull/2740 is pre-requisite to this PR?

cathy-zhou commented:
/retest

jcaamano commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -659,75 +682,132 @@ func getPolicyType(policy *knet.NetworkPolicy) (bool, bool) {
-			errObjs = append(errObjs, pod)
+		// secondary networks
+		on, networkMap, err := util.GetPodNADToNetworkMapping(pod, bnc.NetInfo)
+		if err != nil || !on {
```
Why are we not returning the error here. Add a comment with an explanation please.

npinaeva commented:
Some notes: there is an ACL ExternalIDs change that should be merged before this pr https://github.com/ovn-org/ovn-kubernetes/pull/3334, it will make sure the indexing of the acls is correct between controllers

I think we need to also implement new indices for port groups before this pr, because otherwise syncing port groups for different controllers will be very difficult (and I guess initially we wanted to make sure that all secondary-controller owned objects will have a new set of ids) @jcaamano 

I also see multicast is enabled for secondary networks in the same pr, probably it makes sense to state that in the name, and maybe even split it to at least another commit, if not another pr?

A question about controller names: I see all secondary controllers are named `netInfo.GetNetworkName() + "-network-controller"`, is it guaranteed that these names will always be unique?

One concern I have: it is extremely simple in the current version of the code to mix the objects used by the default and secondary network controllers, which may break the functionality for both. A simple example is `createMulticastAllowPolicy`. If you look at the implementation, it will reference the default network controller port group from
`portGroupName := hashedPortGroup(ns)` in its acls, and the real port group with another network-scoped name will be created in the end with `bnc.buildPortGroup(portGroupName, ns, ports, acls)`. 
I guess there are more examples of that, and maybe if we get the new IDs for all the used objects it will help somewhat, but we need to be extra careful about properly indexing db objects.

jcaamano reviewed (changes requested):
We are not synchronizing multi-network policies on NAD events?

Should the peers of a multinet policy be those on the same network or on the same NAD or all? On `pod_selector_addres_set.go` we are always using the default network for pod peers. For namespace peers we always look in the same network. Something doesn't add up or I am a bit confused.

maiqueb commented on go-controller/pkg/config/config.go:
```diff
@@ -912,6 +913,12 @@ var OVNK8sFeatureFlags = []cli.Flag{
 		Destination: &cliConfig.OVNKubernetesFeature.EnableMultiNetwork,
 		Value:       OVNKubernetesFeature.EnableMultiNetwork,
 	},
+	&cli.BoolFlag{
+		Name:        "enable-multi-networkpolicy",
+		Usage:       "Configure to use multiple networkAttachmentDefinition CRD feature with ovn-kubernetes.",
+		Destination: &cliConfig.OVNKubernetesFeature.EnableMultiNetworkPolicy,
+		Value:       OVNKubernetesFeature.EnableMultiNetworkPolicy,
+	},
```
I would consider @jcaamano 's suggestion, and have only the multi network flag.

IMO, everything becomes simpler.

If we really want to have a separate flag for this, I agree: it should require `EnableMultiNetwork`.

cathy-zhou commented:
> Some notes: there is an ACL ExternalIDs change that should be merged before this pr https://github.com/ovn-org/ovn-kubernetes/pull/3334, it will make sure the indexing of the acls is correct between controllers
> 

sure

> I think we need to also implement new indices for port groups before this pr, because otherwise syncing port groups for different controllers will be very difficult (and I guess initially we wanted to make sure that all secondary-controller owned objects will have a new set of ids) @jcaamano
> 

why it is needed? we have network external-ids to differenciate them. it is like logical_routers/logical_switches/logical_ports

> I also see multicast is enabled for secondary networks in the same pr, probably it makes sense to state that in the name, and maybe even split it to at least another commit, if not another pr?
> 

it is not enabled for secondary networks yet.

> A question about controller names: I see all secondary controllers are named netInfo.GetNetworkName() + "-network-controller", is it guaranteed that these names will always be unique?
> 

I would think so. do you have any specific concerns for this?

> One concern I have: it is extremely simple in the current version of the code to mix the objects used by the default and secondary network controllers, which may break the functionality for both. A simple example is createMulticastAllowPolicy. If you look at the implementation, it will reference the default network controller port group from
> portGroupName := hashedPortGroup(ns) in its acls, and the real port group with another network-scoped name will be created in the end with bnc.buildPortGroup(portGroupName, ns, ports, acls).
> I guess there are more examples of that, and maybe if we get the new IDs for all the used objects it will help somewhat, but we need to be extra careful about properly indexing db objects.

can you point me the code which sets `portGroupName := hashedPortGroup(ns) in its acls`. if that is the case, that is a bug.

as I understand, we need to index the db object only if there is no index for this object type, for example, acls. but for port groups and address-set, name is within the ovndb table, which is index. So I am not sure I understand the needs to introduce yet another index for these kind of objects.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_acl.go:
```diff
@@ -57,10 +57,18 @@ func hashedPortGroup(s string) string {
 // BuildACL should be used to build ACL instead of directly calling libovsdbops.BuildACL.
 // It can properly set and reset log settings for ACL based on ACLLoggingLevels
-func BuildACL(aclName string, priority int, match, action string,
+func (ncni *NetworkControllerNetInfo) BuildACL(aclName string, priority int, match, action string,
```
but this BuildACL function is also called by `gressPolicy`, and that is the reason why I introduce NetworkControllerNetInfo. Do you have any suggestion of the alternative way?

cathy-zhou replied:
do not need this now after acl index changes

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller.go:
```diff
@@ -343,7 +372,7 @@ func (bnc *BaseNetworkController) createNodeLogicalSwitch(nodeName string, hostS
 
 	// multicast is only supported in default network for now
 	if bnc.multicastSupport {
-		err = libovsdbops.AddPortsToPortGroup(bnc.nbClient, types.ClusterRtrPortGroupName, logicalSwitchPort.UUID)
+		err = libovsdbops.AddPortsToPortGroup(bnc.nbClient, bnc.GetNetworkScopedName(types.ClusterRtrPortGroupName), logicalSwitchPort.UUID)
```
multicastSupport is still set to false for secondary networks. I suppose it is not difficult to enable it though. But is it configurable? if so, how to configure it? As part of NAD?

jcaamano replied:
I still find this comment is confusing. Multicast might be disabled for secondary networks but the code is there to support it so I don't think this comment here makes much sense. There is already a comment where multicast is disabled.

But if multicast is not enabled, was there a specific reason to change the code to support it?

cathy-zhou commented on go-controller/pkg/ovn/namespace.go:
```diff
@@ -399,6 +261,12 @@ func (oc *DefaultNetworkController) deleteNamespace(ns *kapi.Namespace) error {
 // ns is the name of the namespace, while namespace is the optional k8s namespace object
 // if no k8s namespace object is provided, this function will attempt to find it via informer cache
 func (oc *DefaultNetworkController) ensureNamespaceLocked(ns string, readOnly bool, namespace *kapi.Namespace) (*namespaceInfo, func(), error) {
+	// special handling of host network namespace. issues/3381
+	//if config.Kubernetes.HostNetworkNamespace != "" && ns == config.Kubernetes.HostNetworkNamespace {
+	//	ips = oc.getAllHostNamespaceAddresses(ns)
+	//}
```
Yes, I understand. But the current upstream code has an issue listed in https://github.com/ovn-org/ovn-kubernetes/issues/3381, result in the hostNamespaceAddress being overridden. That is why I commented out the host network namespace code.

jcaamano replied:
I know. I am suggesting the correct behavior is to expect the management and join LRP IPs are in the address set only, as the test. And that there is probably a bug in the code and in the test. You could update the test to add pods in the namespace and see if it still passes.

cathy-zhou replied:
fixed. needs to fix test case to be able to pass

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,267 @@
+	aclLogging ACLLoggingLevels
+}
+
+func getNamespaceAddrSetDbIDs(namespaceName, controller string) *libovsdbops.DbObjectIDs {
+	return libovsdbops.NewDbObjectIDs(libovsdbops.AddressSetNamespace, controller, map[libovsdbops.ExternalIDKey]string{
+		// namespace has only 1 address set, no additional ids are required
+		libovsdbops.ObjectNameKey: namespaceName,
+	})
+}
```
addressset in the end would have a controller name external-id, which will be per-network.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,267 @@
+
+// This function implements the main body of work of syncNamespaces.
+// Upon failure, it may be invoked multiple times in order to avoid a pod restart.
+func (bnc *BaseNetworkController) syncNamespaces(namespaces []interface{}) error {
```
NAD is processed in parallel, and this function is consulting ovsdb cache with the address-set that only belongs to this network, hopefully that would mitigate the problem.

jcaamano replied:
This function is running a predicate through all the address sets of the address set table. It does so multiple times, per network. We recently became aware that this is not free even if it is the cache. We had a bug were we were spending 14 minutes at startup synchronizing ACLs and 50% of that time was spent running a predicate on the table. 

The point I am trying to make is that we need to be mindful of how we do these things.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_multipolicy.go:
```diff
@@ -0,0 +1,123 @@
+			networkName = a[1]
+			networkNamespace = a[0]
+		}
+		if bnc.HasNAD(util.GetNADName(networkNamespace, networkName)) {
```
in which case when NAD is not processed yet? when ovnkube-master starts, it iterate all the existing NADs first before start all the controller of NADs (see SyncNetworkControllers()). That is to avoid the problem, which could also apply on Pods where nad is defined in their NetworkAttachmentAnnot annotation.

jcaamano replied:
This is used also the a multinet policy is added at runtime, not only on the first synchonization.

So a user could create a nad and a multinet policy at the same time. If we happen to process the multinet policy before the NAD, we check that the multinet policy does not apply and then when we process the NAD we don't go back to check which multinet policies could apply because of that NAD.

cathy-zhou replied:
True, the assumption is that NAD is ready before multi-policy. This seems to be a problem difficult to solve. It means that for every NAD added/deleted, we'd need to trigger an update event for every existing multi-policy.

not sure what is the best way to solve this.

cathy-zhou replied:
based on discussion, this support is deferred

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,268 @@
+	//
+	// You must hold the namespaceInfo's mutex to add/delete dependent policies.
+	// Namespace can take oc.networkPolicies key Lock while holding nsInfo lock, the opposite should never happen.
+	relatedNetworkPolicies map[string]bool
```
this is exisiting upstream code. I didn't change it.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,267 @@
+	validLogLevels := sets.NewString(nbdb.ACLSeverityAlert, nbdb.ACLSeverityWarning, nbdb.ACLSeverityNotice,
+		nbdb.ACLSeverityInfo, nbdb.ACLSeverityDebug, "")
+
+	// Set Deny logging.
+	if validLogLevels.Has(aclLevels.Deny) {
+		nsInfo.aclLogging.Deny = aclLevels.Deny
+	} else {
+		errors = append(errors, fmt.Errorf("disabling deny logging due to invalid deny annotation. "+
+			"%q is not a valid log severity", aclLevels.Deny))
+		nsInfo.aclLogging.Deny = ""
+	}
```
this again is the existing upstream code that I didn't change.

https://github.com/ovn-org/ovn-kubernetes/blob/f3b8b8e7065228011a3515ccda676b2885a2e7cc/go-controller/pkg/ovn/ovn.go#L327

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,267 @@
+		nsInfo.aclLogging.Deny = ""
+	}
+
+	// Set Allow logging.
+	if validLogLevels.Has(aclLevels.Allow) {
+		nsInfo.aclLogging.Allow = aclLevels.Allow
+	} else {
+		errors = append(errors, fmt.Errorf("disabling allow logging due to an invalid allow annotation. "+
+			"%q is not a valid log severity", aclLevels.Allow))
+		nsInfo.aclLogging.Allow = ""
+	}
```
see above

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,267 @@
+// This function implements the main body of work of syncNamespaces.
+// Upon failure, it may be invoked multiple times in order to avoid a pod restart.
+func (bnc *BaseNetworkController) syncNamespaces(namespaces []interface{}) error {
+	expectedNs := make(map[string]bool)
+	for _, nsInterface := range namespaces {
+		ns, ok := nsInterface.(*kapi.Namespace)
+		if !ok {
+			return fmt.Errorf("spurious object in syncNamespaces: %v", nsInterface)
+		}
+		expectedNs[ns.Name] = true
+	}
```
again, this is existing upstream code: https://github.com/ovn-org/ovn-kubernetes/blob/f3b8b8e7065228011a3515ccda676b2885a2e7cc/go-controller/pkg/ovn/namespace.go#L69

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,267 @@
+}
+
+func (bnc *BaseNetworkController) configureNamespaceCommon(nsInfo *namespaceInfo, ns *kapi.Namespace) error {
+	if annotation, ok := ns.Annotations[util.AclLoggingAnnotation]; ok {
```
existing upstream code: https://github.com/ovn-org/ovn-kubernetes/blob/f3b8b8e7065228011a3515ccda676b2885a2e7cc/go-controller/pkg/ovn/namespace.go#L241

jcaamano commented on go-controller/pkg/ovn/gress_policy.go:
```diff
@@ -246,9 +249,11 @@ func (gp *gressPolicy) isEmpty() bool {
 // by the parent NetworkPolicy)
 // buildLocalPodACLs is safe for concurrent use, since it only uses gressPolicy fields that don't change
 // since creation, or are safe for concurrent use like peerVXAddressSets
+// Note that portGroupName is portGroup's name without network scope prefix
```
I am suggesting otherwise, to not rely on `NetworkControllerNetInfo`. 

Another option is to provide the network name and have a `GetNetworkScopedName` free function.

jcaamano commented:
Could we add some basic e2e tests?

npinaeva commented:
>  > A question about controller names: I see all secondary controllers are named netInfo.GetNetworkName() + "-network-controller", is it guaranteed that these names will always be unique?

>I would think so. do you have any specific concerns for this?

Yes, we need to be absolutely sure that the names of controllers are different, because it is a part of objects db index. So let's figure out if/how it is guaranteed.

> > I also see multicast is enabled for secondary networks in the same pr, probably it makes sense to state that in the name, and maybe even split it to at least another commit, if not another pr?

> it is not enabled for secondary networks yet.

I mean it may not be enabled as a flag, but the code for is there, e.g. this pr adds `updateNamespaceForSecondaryNetwork` functino, that calls `multicastUpdateNamespace`. Also the fact that `configureNamespaceCommon` calls `multicastUpdateNamespace` is also confusing

About referencing port groups, looks like it is correct now, I will suggest some changes to make the code more clear about that.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -659,75 +682,132 @@ func getPolicyType(policy *knet.NetworkPolicy) (bool, bool) {
 	klog.Infof("Processing NetworkPolicy %s/%s to have %d local pods...", np.namespace, np.name, len(objs))
-	policyPortUUIDs = make([]string, 0, len(objs))
-	policyPortsToUUIDs = map[string]string{}
+	policyPortUUIDs := []string{}
+	policyPortsToUUIDs := map[string]string{}
```
because for secondary network, there may be more than one policyPortUUID, so the size of policyPortUUIDs is not determined.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -477,32 +488,44 @@ func getARPAllowACLName(ns string) string {
 }
 
+// Note that the hashName argument is the portGroup's Name without network scope prefix
+func (ncni *NetworkControllerNetInfo) buildPortGroup(hashName, name string, ports []*nbdb.LogicalSwitchPort, acls []*nbdb.ACL) *nbdb.PortGroup {
```
I am not sure I understand: this is afunction that all controller use. I don't see issue with prefix after hash, but I don't have strong opinions one way or another. 

The problem for us is that our downstream zone would have issues when upgrading to the latest upstream code if we change the port group name.

cathy-zhou replied:
this is done the same way when we determine the logical router name/port name etc. for secondary networks. Note that I am getting rid of NetworkControllerNetInfo structure now and in both `gressPolicy `and `BaseNetworkController`'s receiver functions needs to call the same function to get port group name.

npinaeva replied:
What I am saying is that calling `GetNetworkScopedName` every time someone needs a real port group name, and adding comments like `Note that the hashName argument is the portGroup's Name without network scope prefix` is error-prone
So I suggest to always build network-prefixed port group name as early as possible by using `getPortGroupName` function, that would be the only function that knows how a port group name is built, and free other functions from that logic

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_acl.go:
```diff
@@ -57,10 +57,18 @@ func hashedPortGroup(s string) string {
 // BuildACL should be used to build ACL instead of directly calling libovsdbops.BuildACL.
 // It can properly set and reset log settings for ACL based on ACLLoggingLevels
-func BuildACL(aclName string, priority int, match, action string,
+func (ncni *NetworkControllerNetInfo) BuildACL(aclName string, priority int, match, action string,
```
@npinaeva when the acl index PR is expected to merge?

npinaeva replied:
by the end of this week, and please reach out to me if you need any help with rebasing, but the code actually should just become simpler (hopefully)

npinaeva commented on go-controller/pkg/ovn/network_conntroller_policy_event_handler.go:
```diff
@@ -0,0 +1,356 @@
+
+// event handlers handles policy related events
+type networkControllerPolicyEventHandler struct {
+	baseHandler     baseNetworkControllerEventHandler
```
I guess `baseHandler` is not really required here, because `networkControllerPolicyEventHandler` handles all netpol-related events, so you can just move these parts frof `baseNetworkControllerEventHandler`?

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_multipolicy.go:
```diff
@@ -0,0 +1,123 @@
+	return false
+}
+
+func convertMultiNetPolicyToNetPolicy(mpolicy *multinetworkpolicyapi.MultiNetworkPolicy) *knet.NetworkPolicy {
```
sorry, do not understand your comments. I agree this function is weird, it is only here to help reuse and not changing too much the existing network policy code.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -477,32 +488,44 @@ func getARPAllowACLName(ns string) string {
+func (ncni *NetworkControllerNetInfo) buildPortGroup(hashName, name string, ports []*nbdb.LogicalSwitchPort, acls []*nbdb.ACL) *nbdb.PortGroup {
+	var externalIds map[string]string
+	pgName := ncni.GetNetworkScopedName(hashName)
+	externalIds = map[string]string{"name": name}
```
I don't think we need. we have network specific external-ids

jcaamano replied:
Not because we need it, it just because that is the PG name and clearly there shouldn't be multiple PGs with the same name, just to be consistent.

npinaeva replied:
i had the same question, but we don't really use externalIDs for port groups, and if we change the value it will be more difficult to sync port groups to the new ids in the future (now we will have the same "name" for any controller type, and we can get owner controller from types.NetworkExternalID without extra parsing)

jcaamano replied:
hmm okay!

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -1660,28 +1741,66 @@ func (oc *DefaultNetworkController) addHairpinAllowACL() error {
+
+// WatchNetworkPolicy starts the watching of multinetworkpolicy resource and calls
+// back the appropriate handler logic
+func (bnc *BaseNetworkController) WatchNetworkPolicy() error {
```
why? they are doing the exactly the same thing.

npinaeva replied:
My general logic here is if
1. the function is called explicitly from default and secondary controllers, and not from `BaseNetworkController` methods
2. and it has different implementations for default and secondary controllers
then it should be 2 different functions.

in this case, `WatchNetworkPolicy` is called from the `Run` method of every controller explicitly, and has this `bnc.IsSecondary()` branching, so it may be 2 separate function for default and secondary controller

Also, `policyHandler` field is only required for secondary controllers, so it may also be moved from the `BaseNetworkController`. Then `WatchNetworkPolicy` for default network controller will stay exactly as before
```
func (oc *DefaultNetworkController) WatchNetworkPolicy() error {
	_, err := oc.retryNetworkPolicies.WatchResource()
	return err
}
```
and secondary controller `WatchNetworkPolicy` method may add all these extra checks and fields

cathy-zhou replied:
ok

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -1660,28 +1741,66 @@ func (oc *DefaultNetworkController) addHairpinAllowACL() error {
+	asPredicate := func(item *nbdb.AddressSet) bool {
+		return item.ExternalIDs[libovsdbops.OwnerControllerKey.String()] == controllerName
+	}
+	ops, err = libovsdbops.DeleteAddressSetsWithPredicateOps(nbClient, ops, asPredicate)
```
all addresssets belongs to this network.

npinaeva replied:
but it actually deletes all address sets and port groups (including multicast port group), so i guess the name is not really correct, maybe you can just put the contents of this function inline to the secondary controller cleanup?
Also, can we please move this function to the secondary controller code, because it is not used by the default network controller.

cathy-zhou replied:
This function is shared by all secondary networks, so I'd rather not to embed them in all secondary network controller cleanup code. any suggestion on naming? multicast support is kind of policy (where or not allow multicast), so I personally think that function name is ok.

cathy-zhou commented on go-controller/pkg/ovn/gress_policy.go:
```diff
@@ -68,16 +70,17 @@ func (pp *portPolicy) getL4Match() (string, error) {
-		peerV6AddressSets: &sync.Map{},
-		portPolicies:      make([]*portPolicy, 0),
+		NetworkControllerNetInfo: NetworkControllerNetInfo{NetInfo: netInfo},
+		controllerName:           controllerName,
```
only when acls are also indexed with controllerName. This will be changed after your acl index PR merged.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_secondary.go:
```diff
@@ -282,3 +308,154 @@ func (bsnc *BaseSecondaryNetworkController) syncPodsForSecondaryNetwork(pods []i
+func (bsnc *BaseSecondaryNetworkController) updateNamespaceForSecondaryNetwork(old, newer *kapi.Namespace) error {
+	var errors []error
+	klog.Infof("[%s] updating namespace for network %s", old.Name, bsnc.GetNetworkName())
+
+	nsInfo, nsUnlock := bsnc.getNamespaceLocked(old.Name, false)
+	if nsInfo == nil {
+		klog.Warningf("Update event for unknown namespace %q", old.Name)
+		return nil
+	}
+	defer nsUnlock()
+
+	aclAnnotation := newer.Annotations[util.AclLoggingAnnotation]
+	oldACLAnnotation := old.Annotations[util.AclLoggingAnnotation]
+	// support for ACL logging update, if new annotation is empty, make sure we propagate new setting
+	if aclAnnotation != oldACLAnnotation {
+		if err := bsnc.updateNamespaceAclLogging(old.Name, aclAnnotation, nsInfo); err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	if err := bsnc.multicastUpdateNamespace(newer, nsInfo); err != nil {
+		errors = append(errors, err)
+	}
+	return kerrors.NewAggregate(errors)
+}
+
+func (bsnc *BaseSecondaryNetworkController) deleteNamespace4SecondaryNetwork(ns *kapi.Namespace) error {
+	klog.Infof("[%s] deleting namespace for network %s", ns.Name, bsnc.GetNetworkName())
+
+	nsInfo := bsnc.deleteNamespaceLocked(ns.Name)
+	if nsInfo == nil {
+		return nil
+	}
+	defer nsInfo.Unlock()
+
+	if err := bsnc.multicastDeleteNamespace(ns, nsInfo); err != nil {
+		return fmt.Errorf("failed to delete multicast nameosace error %v", err)
+	}
+	return nil
+}
```
honestly, I am not sure what is the better way to share the code. Especially between ensureNamespaceLocked() and ensureNamespaceLockedForSecondaryNetwork()

jcaamano replied:
That's one option. 

The other option is to split ensureNamespaceLocked because the first 60 lines of that method are practically the same than ensureNamespaceLockedForSecondaryNetwork.

cathy-zhou replied:
I don't see how to split it based on how locking is done today.

cathy-zhou replied:
> what will we be doing differently between the default network controller and the secondary network controller with IPAM?
> 
> The way I see it, the big difference will come for an IPAMless secondary controller, which will not have to track namespaces / pods labels and keep IP based address sets in sync.

then for ipamless secondary controller, we do not need watch for namespaces at all. I've made the change.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -511,50 +534,50 @@ func (oc *DefaultNetworkController) addPolicyToDefaultPortGroups(np *networkPoli
 // must be called with defaultDenyPortGroups lock
-func (oc *DefaultNetworkController) createDefaultDenyPGAndACLs(namespace, policy string, aclLogging *ACLLoggingLevels) error {
+func (bnc *BaseNetworkController) createDefaultDenyPGAndACLs(namespace, policy string, aclLogging *ACLLoggingLevels) error {
 	ingressPGName := defaultDenyPortGroupName(namespace, ingressDefaultDenySuffix)
```
note that this is the group name without network prefix.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -1129,27 +1205,27 @@ func (oc *DefaultNetworkController) createNetworkPolicy(policy *knet.NetworkPoli
 
-		pg := libovsdbops.BuildPortGroup(np.portGroupName, readableGroupName, nil, acls)
-		ops, err = libovsdbops.CreateOrUpdatePortGroupsOps(oc.nbClient, ops, pg)
+		pg := bnc.buildPortGroup(np.portGroupName, readableGroupName, nil, acls)
```
I'd rather leave np.portGroupName as the name without network prefix. Then add the network scope prefix at last when we needs to build the logical entities. So it is easy to determine the form of the port group name (either prefix+hash(name), or hash(prefix+name))

npinaeva replied:
`portGroupName` in networkPolicy object is supposed to be used as portGroup.Name, e.g. in acl.Match, or other db ops to index port group. All these operation don't care about the format of that name, they just need portGroup.Name as it is stored in the db. So I am not sure why do you think we need to "determine the form of the port group name"? AFAICS every usage of `np.portGroupName` ends up with `GetNetworkScopedName(np.portGroupName)`

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -1638,16 +1715,20 @@ func getStaleNetpolAddrSetDbIDs(policyNamespace, policyName, policyType, idx, co
+func (bnc *BaseNetworkController) addHairpinAllowACL() error {
 	var v4Match, v6Match, match string
+
+	if bnc.IsSecondary() {
```
that means we'd need another set of networkControllerPolicyEventHandler only to handle policy related events for default network ,and they are all same (copy and paste) execpt this one. I'd prefer not to do that.

npinaeva replied:
Yeah I understand, what I mean is `addHairpinAllowACL` is called from `syncNetworkPoliciesCommon`, but it may be only called from `syncNetworkPolicies` instead, because it does nothing for `syncMultiNetworkPolicies`, so there is no sense in including it to `syncNetworkPoliciesCommon`?

cathy-zhou replied:
ok, that needs to move retries of policy from networkControllerPolicyEventHandler to default/layer3/baseLayer3/NetworkControllerEventHandler. I've made the change.

cathy-zhou commented on go-controller/pkg/ovn/gress_policy.go:
```diff
@@ -246,9 +249,11 @@ func (gp *gressPolicy) isEmpty() bool {
 // by the parent NetworkPolicy)
 // buildLocalPodACLs is safe for concurrent use, since it only uses gressPolicy fields that don't change
 // since creation, or are safe for concurrent use like peerVXAddressSets
+// Note that portGroupName is portGroup's name without network scope prefix
```
see above. regardless the final decision regarding the form of portgroup name, I'd rather not to change this. I don't the need to do it. Basically, the idea is to defer naming of the portGroup at the end before ovn logical entity is created.

maiqueb reviewed (commented):
some more questions.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_secondary.go:
```diff
@@ -282,3 +308,154 @@ func (bsnc *BaseSecondaryNetworkController) syncPodsForSecondaryNetwork(pods []i
+		klog.Infof("[%s] adding namespace took %v for network %s", ns.Name, time.Since(start), bsnc.GetNetworkName())
+	}()
+
+	_, nsUnlock, err := bsnc.ensureNamespaceLockedForSecondaryNetwork(ns.Name, false, ns)
```
this is a variants of ensureNamespaceLocked() function for default networks. It does the same thing except the things that does not exist in secondary networks.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_secondary.go:
```diff
@@ -282,3 +308,154 @@ func (bsnc *BaseSecondaryNetworkController) syncPodsForSecondaryNetwork(pods []i
+		}
+		// we are creating nsInfo and going to set it in namespaces map
+		// so safe to hold the lock while we create and add it
+		defer bsnc.namespacesMutex.Unlock()
```
because depends on if nsInfoExisted, the unlock happens in different place. This is the same as default network ensureNamespaceLocked().

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_secondary.go:
```diff
@@ -47,6 +51,13 @@ func (bsnc *BaseSecondaryNetworkController) AddSecondaryNetworkResourceCommon(ob
 		}
 		return bsnc.ensurePodForSecondaryNetwork(pod, true)
 
+	case factory.NamespaceType:
+		ns, ok := obj.(*kapi.Namespace)
+		if !ok {
+			return fmt.Errorf("could not cast %T object to *kapi.Namespace", obj)
+		}
+		return bsnc.AddNamespaceForSecondaryNetwork(ns)
+
```
didn't think this through yet. But I don't think you need namespace related handling if policy is not based on namespace/pod selector. If so, you don't need to watch for namespace at all. you could simply direct return in WatchNamespaces() without calling WatchResource()

maiqueb replied:
@cathy-zhou feel free to keep working exclusively the IPAM case in this PR.

I can tackle the non-IPAM case as a follow-up. 

I think this would help keep this PR "small", otherwise, it'll bloat even more.

npinaeva commented on go-controller/pkg/ovn/external_ids_syncer/acl/acl_sync_test.go:
```diff
@@ -69,6 +69,11 @@ func joinACLName(substrings ...string) string {
 	return strings.Join(substrings, "_")
 }
 
+func buildPortGroup(hashName, name string, ports []*nbdb.LogicalSwitchPort, acls []*nbdb.ACL) *nbdb.PortGroup {
```
the sync should never be updated, because the code only works with the previous version of ids, so I think you can just revert changes of this file and it should work

cathy-zhou replied:
I made this change because it does not compile

npinaeva commented on go-controller/pkg/ovn/network_conntroller_policy_event_handler.go:
```diff
@@ -0,0 +1,356 @@
+	"k8s.io/klog/v2"
+)
+
+func (bnc *BaseNetworkController) initRetryFramework() {
```
same logic I explained above: this function is directly called from the default and secondary controllers, so you can just call the right function from there

cathy-zhou replied:
ok

cathy-zhou commented on go-controller/pkg/ovn/gress_policy.go:
```diff
@@ -11,13 +11,15 @@ import (
 type gressPolicy struct {
-	controllerName  string
+	controllerName string
+	NetworkControllerNetInfo
```
removed.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -659,75 +682,132 @@ func getPolicyType(policy *knet.NetworkPolicy) (bool, bool) {
-func (oc *DefaultNetworkController) getNewLocalPolicyPorts(np *networkPolicy,
-	objs ...interface{}) (policyPortsToUUIDs map[string]string, policyPortUUIDs []string, errObjs []interface{}) {
+// if there are problems with fetching port info from logicalPortCache, error will be added to returned error array.
+func (bnc *BaseNetworkController) getNewLocalPolicyPorts(np *networkPolicy,
```
the reason why I made the current change is that for default network, GetPodNADToNetworkMapping() does not return the "default" NADName as expected. It may return real NAD name that is annoated in `v1.multus-cni.io/default-network`. That is why we usually need to have a xxxForNAD() function (see as addLogicalPortToNetwork())

cathy-zhou replied:
hmm, I don't find the code less clear. Also we'd like to make use of the same GetPodNADToNetworkMapping() anyway (not to invent another function doing things similiar), so I don't see how performance could be different.

npinaeva replied:
we have at least 2 duplicated code parts
```
on, networkMap, err := util.GetPodNADToNetworkMapping(pod, bnc.NetInfo)
		if err != nil || !on {
			if err != nil {
				// if we are not able to determine if this pod is on this network continue processing the remain pods
				// anyway, hopefully the pod will be handled in a future pod update event.
				klog.Warningf("Failed to determine if pod %s/%s is on network %s: %v",
					pod.Namespace, pod.Name, bnc.GetNetworkName(), err)
			}
			continue
		}
```
is called from `getNewLocalPolicyPorts` and `getExistingLocalPolicyPorts`
So I would like to have a single function that will do this part, and if you just add return "" for default network controller, then we can remove the second duplication which is:

```
portInfo, err := bnc.getNewLocalPolicyPortsForNAD(np, pod, types.DefaultNetworkName)
			if err != nil {
				errs = append(errs, err)
			} else if portInfo != nil {
				policyPortUUIDs = append(policyPortUUIDs, portInfo.uuid)
				policyPortsToUUIDs[portInfo.name] = portInfo.uuid
			}
			continue
```
is called in `getNewLocalPolicyPorts` twice: separately for default and secondary controllers.
So you can keep using `GetPodNADToNetworkMapping` from `bnc.getNADNames` and still have the code as I suggested before ^.

also
```
if !bnc.IsSecondary() {
		logicalPortName = util.GetLogicalPortName(pod.Namespace, pod.Name)
	} else {
		logicalPortName = util.GetSecondaryNetworkLogicalPortName(pod.Namespace, pod.Name, nadName)
	}
```
can be replaced with `bnc.GetLogicalPortName(pod, nadName)`, and let us have no extra `if !bnc.IsSecondary()` in the handlers

cathy-zhou commented:
@maiqueb @npinaeva @jcaamano I've fixed all the issues execpt the portgroup name related, which I will do it after we made the the agreement on its form. Please take a look and let me know if I missed anything.

I will also start to work on unit tests as well.

cathy-zhou commented:
@npinaeva @jcaamano @maiqueb Please take a look of the latest changes. Thanks!

npinaeva commented on go-controller/pkg/ovn/pod_selector_address_set.go:
```diff
@@ -136,39 +138,42 @@ func (oc *DefaultNetworkController) EnsurePodSelectorAddressSet(podSelector, nam
 			podSelector:       psas.podSelector,
 			namespaceSelector: psas.namespaceSelector,
 			namespace:         psas.namespace,
+			ipv4Mode:          ipv4Mode,
+			ipv6Mode:          ipv6Mode,
```
these variables are only used to build the slice of the right size, I would say it is better to just remove them from the object, and use empty constructor

cathy-zhou replied:
I am not sure I understand your suggestion. Can you please be more specific?

npinaeva replied:
it is different now since we have `netInfo` also. Are we going to be able to get ipMode from it as @jcaamano suggested?

cathy-zhou replied:
ipMode is not part of netInfo, it is attribute of netConfInfo. But @jcaamano is make a change to merge them together. but not in this PR.

npinaeva commented on go-controller/pkg/ovn/pod_selector_address_set.go:
```diff
@@ -335,7 +343,7 @@ func (handlerInfo *PodSelectorAddrSetHandlerInfo) addPods(pods ...*v1.Pod) error
 
 	podIPFactor := 1
-	if config.IPv4Mode && config.IPv6Mode {
+	if handlerInfo.ipv4Mode && handlerInfo.ipv6Mode {
```
this function together with `deletePod` call 
`util.GetPodIPsOfNetwork(pod, &util.DefaultNetInfo{})`
is it intended to use default network here?

maiqueb replied:
good question - I think it should be using the network to which the address set corresponds.

 I'm running this PR locally and I think the pod selector addresses set features IPs from the wrong network:
```
_uuid               : a914ea2b-14ce-4e34-a639-bb0a1ff16c33
addresses           : ["192.168.100.1", "192.168.100.2", "192.168.100.3"]
external_ids        : {ip-family=v4, "k8s.ovn.org/id"="tenant-blue-network-controller:Namespace:default:v4", "k8s.ovn.org/name"=default, "k8s.ovn.org/owner-controller"=tenant-blue-network-controller, "k8s.ovn.org/owner-type"=Namespace}
name                : a9836740957047443908

# I think this address set should have an IP address of the secondary network, not from the primary
_uuid               : 7ab6702f-a699-470d-818d-a15aa9904934
addresses           : ["10.244.0.14"]
external_ids        : {ip-family=v4, "k8s.ovn.org/id"="tenant-blue-network-controller:PodSelector:default_LS{ML:{app: allowed-client,},}:v4", "k8s.ovn.org/name"="default_LS{ML:{app: allowed-client,},}", "k8s.ovn.org/owner-controller"=tenant-blue-network-controller, "k8s.ovn.org/owner-type"=PodSelector}
name                : a1493214346180864325
```

Here's the net-attach-def / policy I'm using:
```yaml
---
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"k8s.cni.cncf.io/v1","kind":"NetworkAttachmentDefinition","metadata":{"annotations":{},"name":"tenant-blue","namespace":"default"},"spec":{"config":"{ \"cniVersion\": \"0.4.0\", \"name\": \"tenant-blue\", \"netAttachDefName\": \"default/tenant-blue\", \"topology\": \"layer2\", \"type\": \"ovn-k8s-cni-overlay\", \"subnets\": \"192.168.100.0/24\", \"logFile\": \"/var/log/ovn-kubernetes/ovn-k8s-cni-overlay.log\", \"logLevel\": \"5\", \"logfile-maxsize\": 100, \"logfile-maxbackups\": 5, \"logfile-maxage\": 5 }"}}
  creationTimestamp: "2023-04-12T10:41:26Z"
  generation: 1
  name: tenant-blue
  namespace: default
  resourceVersion: "5904"
  uid: d1e70f5f-4ab2-4d5d-900a-9988746a5f6b
spec:
  config: '{ "cniVersion": "0.4.0", "name": "tenant-blue", "netAttachDefName": "default/tenant-blue",
    "topology": "layer2", "type": "ovn-k8s-cni-overlay", "subnets": "192.168.100.0/24",
    "logFile": "/var/log/ovn-kubernetes/ovn-k8s-cni-overlay.log", "logLevel": "5",
    "logfile-maxsize": 100, "logfile-maxbackups": 5, "logfile-maxage": 5 }'
---
apiVersion: k8s.cni.cncf.io/v1beta1
kind: MultiNetworkPolicy
metadata:
  annotations:
    k8s.v1.cni.cncf.io/policy-for: tenant-blue
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"k8s.cni.cncf.io/v1beta1","kind":"MultiNetworkPolicy","metadata":{"annotations":{"k8s.v1.cni.cncf.io/policy-for":"tenant-blue"},"name":"allow-ports-same-ns","namespace":"default"},"spec":{"ingress":[{"from":[{"podSelector":{"matchLabels":{"app":"allowed-client"}}}]}],"podSelector":{"matchLabels":{"app":"stuff-doer"}}}}
  creationTimestamp: "2023-04-12T10:41:26Z"
  generation: 1
  name: allow-ports-same-ns
  namespace: default
  resourceVersion: "5905"
  uid: bf691987-53fd-4bc8-bddd-07306231b56f
spec:
  ingress:
  - from:
    - podSelector:
        matchLabels:
          app: allowed-client
  podSelector:
    matchLabels:
      app: stuff-doer
```

View from a pod:
```bash
kubectl exec pod1 -- ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0@if30: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP group default 
    link/ether 0a:58:0a:f4:00:0e brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.244.0.14/24 brd 10.244.0.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::858:aff:fef4:e/64 scope link 
       valid_lft forever preferred_lft forever
3: net1@if32: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP group default 
    link/ether 0a:58:c0:a8:64:01 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 192.168.100.1/24 brd 192.168.100.255 scope global net1
       valid_lft forever preferred_lft forever
    inet6 fe80::858:c0ff:fea8:6401/64 scope link 
       valid_lft forever preferred_lft forever

```

maiqueb replied:
@cathy-zhou

cathy-zhou replied:
> this function together with deletePod call
> util.GetPodIPsOfNetwork(pod, &util.DefaultNetInfo{})
> is it intended to use default network here?

npinaeva commented on go-controller/pkg/ovn/pod_selector_address_set.go:
```diff
@@ -731,3 +739,31 @@ func deleteAddrSetsWithoutACLRef(predicateIDs *libovsdbops.DbObjectIDs,
 	return nil
 }
+
+func (bnc *BaseNetworkController) syncAddressSets() error {
```
If this is only used by the default network controller, it should be moved to `default_network_controller.go`

cathy-zhou replied:
It would be needed for secondary networks as well for downstream.

cathy-zhou replied:
I think it should work with secondary networks with minor changes (just search for no-owner object and external-ids:network_name belongs to a certain network.

npinaeva reviewed (commented):
Some more comments, please check some of the changes that just moved some functions in the same file (I pointed out some of those), let's revert them to not make extra changes by just moving thing around.
About port group naming: most of my comments are independent of the decision about the name itself, if you create a function `func (ncni *NetworkControllerNetInfo) getPortGroupName(name string) string` that I described in one of the comments, it will be the only place to change when we come up with the final decision.

npinaeva commented on go-controller/pkg/ovn/base_network_controller_acl.go:
```diff
@@ -119,7 +119,8 @@ func BuildACL(dbIDs *libovsdbops.DbObjectIDs, priority int, match, action string
 }
 
-func getACLMatch(portGroupName, match string, aclDir aclDirection) string {
+// Note that the portGroupName argument is the portGroup's Name without network scope prefix
```
I guess this comment is outdated now

maiqueb reviewed (changes requested):
Some more comments: I've spotted a nil pointer while trying out the PR, and I also think we're storing the IP address of the default network on the address set for the secondary network.

maiqueb commented:
@cathy-zhou can you update this PR's description w/ a multi-network policy + net-attach-def you know works ?

My e2e tests so far all seem to fail. I can push a PR with them if that helps you.

cathy-zhou commented on go-controller/pkg/ovn/address_set/fake_address_set.go:
```diff
@@ -210,6 +228,14 @@ func (f *FakeAddressSetFactory) getDbIDsFromNsNameOrDbIDs(dbIDsOrNsName any) *li
 
+// ExpectAddressSetHasIPs ensure address set has the given set of ips.
+// Address set is identified by dbIDsOrNsName, which may be a namespace name (string) or a *libovsdbops.DbObjectIDs.
+func (f *FakeAddressSetFactory) ExpectAddressSetHasIPs(dbIDsOrNsName any, ips []string) {
```
ExpectAddressSetWithIPs is checking the address-set has the same ips as the input argument, ExpectAddressSetHasIPs is to check the address-set has (but may be more ips than ) the ips in the input argument.

see the comment where this new function is called:

			// check the namespace and ensure the address set being created with the right set of IPs in it.
			// Note that NodeMgmtPortIP may or may not in the hostNetworkNamespace's address set, depends on
			// if node1's subnet annotation already exists in the informer cache or not when hostNetwork
			// namespace is handled in WatchNamespaces.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller.go:
```diff
@@ -54,11 +55,14 @@ type CommonNetworkControllerInfo struct {
 	// Supports OVN Template Load Balancers?
 	svcTemplateSupport bool
+	// Is ACL logging enabled while configuring meters?
+	aclLoggingEnabled bool
```
as I said, I just followed the existing behavior for default network.

jcaamano commented on go-controller/pkg/ovn/multipolicy_test.go:
```diff
@@ -0,0 +1,543 @@
+		var switchName string
+		switch ocInfo.topology {
+		case ovntypes.Layer3Topology:
+			l3Controller := ocInfo.NetworkController.(*SecondaryLayer3NetworkController)
```
we might as well embed or make NetInfo retrievable from the NetwrokController interface

cathy-zhou replied:
how can that help. We'd still need its embeded BaseSecondaryNetworkController to watch for namepsace, pod and multi-network-policy.

jcaamano replied:
I was just looking at this method. 

But indeed, if there is multiple reasons to do this, then I think it is better to encapsulate these type of things in `FakeOVN` itself. Otherwise every other test suite will need to replicate this code.

For example, this `getExpectedDataPodsAndSwitchesForSecondaryNetwork` looks like something that any test suite that deals with secondary networks will need.

Also, FakeOVN could just offer a method `getNetworkController(networkName)` that returns a test specific interface with the methods needed like `TopologyType` or `WatchNamespaces`. If not all the network controllers implement all methods it could just be an adapter that does nothing for the methods not implemented.

cathy-zhou replied:
I've changed the FakeOVN.secondaryControllers to be map of secondary network controller information, which includes the BaseSecondaryNetworkController. It should be easily used by any other consumer if needed.

jcaamano reviewed (changes requested):
I resolved the comments that were addressed and added some new ones.

Comments that were not addressed or not answered:

- https://github.com/ovn-org/ovn-kubernetes/pull/3382#discussion_r1150346243
- https://github.com/ovn-org/ovn-kubernetes/pull/3382#pullrequestreview-1362803528
Should the peers of a multinet policy be those on the same network or on the same NAD or all? On pod_selector_addres_set.go we are always using the default network for pod peers. For namespace peers we always look in the same network. Something doesn't add up or I am a bit confused.
- https://github.com/ovn-org/ovn-kubernetes/pull/3382#discussion_r1152174738
- https://github.com/ovn-org/ovn-kubernetes/pull/3382#discussion_r1152184929

@cathy-zhou if you could please rework on additional commits or give an overview of the changes not related to the comments in each push we may save a bit of time reviewing. Thank you!

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,353 @@
+// Note that controller's namespacesMutex may or may not be hold, which is indicated by returned
+// needUnlockNamespace boolean value. Caller needs to release it after configure nsInfo.
+func (bnc *BaseNetworkController) ensureNamespaceLockedCommon(ns string, readOnly bool, namespace *kapi.Namespace,
+	ips []net.IP) (*kapi.Namespace, *namespaceInfo, func(), bool, error) {
```
what about the case when nsInfoExisted is false, the function returns success, but the caller cannot call bnc.namespacesMutex.Unlock().

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -519,74 +515,133 @@ func getPolicyType(policy *knet.NetworkPolicy) (bool, bool) {
+		on, networkMap, err := util.GetPodNADToNetworkMapping(pod, bnc.NetInfo)
+		if err != nil || !on {
+			if err != nil {
+				// continue to handle other pods even if this pod has invalid Network Attachment Selection Annotation
```
we should not retry, it is just a configuration error. GetPodNADToNetworkMapping returns error only when k8s.v1.cni.cncf.io/networks annotation is incorrectly configured.

jcaamano replied:
Then please add that reasoning in the comment. I guess that if the user fixes the annotation, we will handle a pod update event which will consider the pod for the policy. Please add these two comments
```
		// skip pods that are not attached to this network
		on, networkMap, err := util.GetPodNADToNetworkMapping(pod, bnc.NetInfo)
		if err != nil || !on {
			if err != nil {
				// if we are not able to determine if a pod is on this network continue processing the remain pods
				// anyway, hopefully this pod will be handled in a future pod update event.
				klog.Warningf("Failed to determine if pod %s/%s needs to be plumb interface on network %s: %v",
					pod.Namespace, pod.Name, bnc.GetNetworkName(), err)
			}
			continue
		}
```

cathy-zhou commented on go-controller/pkg/ovn/gress_policy.go:
```diff
@@ -91,6 +99,8 @@ func newGressPolicy(policyType knet.PolicyType, idx int, namespace, name, contro
 		peerV6AddressSets: &sync.Map{},
 		portPolicies:      make([]*portPolicy, 0),
 		isNetPolStateless: isNetPolStateless,
+		ipv4Mode:          ipv4Mode,
+		ipv6Mode:          ipv6Mode,
```
it is not part of NetInfo, it is NetConfInfo()

jcaamano replied:
Ah right. I am merging those two, could you PTAL here: https://github.com/ovn-org/ovn-kubernetes/pull/3527

cathy-zhou replied:
will do

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_secondary.go:
```diff
@@ -282,3 +380,149 @@ func (bsnc *BaseSecondaryNetworkController) syncPodsForSecondaryNetwork(pods []i
+	}
+
+	// if this network does not have ipam, network policy is not supported.
+	if !bsnc.doesNetworkRequireIPAM() {
+		klog.Infof("Network policy is not supported on network %s", bsnc.GetNetworkName())
+		return nil
+	}
```
it is not supported for now. you can remove it after you add the support for network policies on ipamless networks.

cathy-zhou commented on go-controller/pkg/ovn/base_secondary_layer2_network_controller.go:
```diff
@@ -163,9 +199,15 @@ func (oc *BaseSecondaryLayer2NetworkController) Stop() {
 	if oc.podHandler != nil {
 		oc.watchFactory.RemovePodHandler(oc.podHandler)
 	}
+	if oc.namespaceHandler != nil {
```
namespaceHandler  will be nil if it namespace watcher is not added, right?

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller.go:
```diff
@@ -103,9 +113,24 @@ type BaseNetworkController struct {
+	// make sure to keep this order to avoid deadlocks
+	sharedNetpolPortGroups *syncmap.SyncMap[*defaultDenyPortGroups]
+
+	podSelectorAddressSets *syncmap.SyncMap[*PodSelectorAddressSet]
```
good cache. Thank you!

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -519,74 +515,133 @@ func getPolicyType(policy *knet.NetworkPolicy) (bool, bool) {
+		on, networkMap, err := util.GetPodNADToNetworkMapping(pod, bnc.NetInfo)
+		if err != nil || !on {
+			if err != nil {
+				// continue to handle other pods even if this pod has invalid Network Attachment Selection Annotation
```
same as below

cathy-zhou commented on go-controller/pkg/ovn/pod_selector_address_set.go:
```diff
@@ -731,3 +739,31 @@ func deleteAddrSetsWithoutACLRef(predicateIDs *libovsdbops.DbObjectIDs,
 	return nil
 }
+
+func (bnc *BaseNetworkController) syncAddressSets() error {
```
I made it as a receiver function of BaseNetworkController, because it will be needed for all networks downstream.

jcaamano replied:
Even if you keep it here now, someone in the future is probably bound to move it to a more appropriate place. How are you going to avoid that? Can't this be moved in your downstream, along with the other differences you must be already handling.

Anyway, we need a more appropriate name for this method because is is also sync'ing ACLs.

cathy-zhou replied:
ok, I will move it to default_network_controller.go and change the name to syncAddressSetsAndAcls()

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_acl.go:
```diff
@@ -106,8 +116,8 @@ func getACLMatch(portGroupName, match string, aclT aclType) string {
 // nsInfo will be locked (and unlocked at the end) for given namespace if it exists.
-func (oc *DefaultNetworkController) GetNamespaceACLLogging(ns string) *ACLLoggingLevels {
-	nsInfo, nsUnlock := oc.getNamespaceLocked(ns, true)
+func (bnc *BaseNetworkController) GetNamespaceACLLogging(ns string) *ACLLoggingLevels {
```
we now have a getACLMatch() which is also receiver function of BaseNetworkController

npinaeva replied:
so can we now do what is suggested in first comment?

cathy-zhou commented on go-controller/pkg/ovn/address_set/fake_address_set.go:
```diff
@@ -155,6 +155,24 @@ func (f *FakeAddressSetFactory) removeAddressSet(name string) {
+// expectAddressSetHasIPs ensures the named address set has the given set of IPs
+func (f *FakeAddressSetFactory) expectAddressSetHasIPs(g gomega.Gomega, dbIDs *libovsdbops.DbObjectIDs, ips []string) {
+	as := f.getAddressSet(dbIDs)
+	gomega.Expect(as).ToNot(gomega.BeNil(), fmt.Sprintf("expected address set %s to exist", dbIDs.String()))
```
I am not sure I understand your suggestion. The problem here is that IP addresses in FakeAddressSetFactory is undetermined, it may have both lrpIP and NodeMgmtPortIP, or just lrpIP

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_multipolicy.go:
```diff
@@ -0,0 +1,124 @@
+	policyForAnnot = strings.ReplaceAll(policyForAnnot, " ", "")
+	policyForNetworks := strings.Split(policyForAnnot, ",")
+	for _, networkName := range policyForNetworks {
+		networkNamespace := mpolicy.Namespace
```
it can change if in PolicyForAnnotation annotation, it specifies the namespace explicitly.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	// Valid log levels are the various preestablished levels or the empty string.
+	validLogLevels := sets.NewString(nbdb.ACLSeverityAlert, nbdb.ACLSeverityWarning, nbdb.ACLSeverityNotice,
+		nbdb.ACLSeverityInfo, nbdb.ACLSeverityDebug, "")
```
this is upstream code, I didn't change it.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,353 @@
+	return apierrors.NewAggregate(errors)
+}
+
+// This function implements the main body of work of syncNamespaces.
+// Upon failure, it may be invoked multiple times in order to avoid a pod restart.
```
this is existing function/comments, I didn't make any changes.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_namespace.go:
```diff
@@ -0,0 +1,353 @@
+
+	err := bnc.addressSetFactory.ProcessEachAddressSet(bnc.controllerName, libovsdbops.AddressSetNamespace,
+		func(dbIDs *libovsdbops.DbObjectIDs) error {
+			if !expectedNs[dbIDs.GetObjectID(libovsdbops.ObjectNameKey)] {
```
but this is upstream code that I didn't change.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_secondary.go:
```diff
@@ -282,3 +380,149 @@ func (bsnc *BaseSecondaryNetworkController) syncPodsForSecondaryNetwork(pods []i
+func (bsnc *BaseSecondaryNetworkController) ensureNamespaceLockedForSecondaryNetwork(ns string, readOnly bool, namespace *kapi.Namespace) (*namespaceInfo, func(), error) {
+	ips := bsnc.getAllNamespacePodAddresses(ns)
+
+	namespace, nsInfo, unlockFunc, needUnlockNamespace, err := bsnc.ensureNamespaceLockedCommon(ns, readOnly, namespace, ips)
+	if needUnlockNamespace {
+		defer bsnc.namespacesMutex.Unlock()
+	}
```
it is different unlock function.

cathy-zhou commented on go-controller/pkg/ovn/base_secondary_layer2_network_controller.go:
```diff
@@ -194,10 +241,21 @@ func (oc *BaseSecondaryLayer2NetworkController) Run() error {
 
+	// WatchNamespaces() should be started first because it has no other
+	// dependencies, and WatchNodes() depends on it
+	if err := oc.WatchNamespaces(); err != nil {
+		return err
+	}
```
it is checked inside the WatchNamespaces() function.

cathy-zhou commented on go-controller/pkg/ovn/namespace_test.go:
```diff
@@ -295,7 +295,16 @@ var _ = ginkgo.Describe("OVN Namespace Operations", func() {
 			err = fakeOvn.controller.WatchNamespaces()
 			gomega.Expect(err).NotTo(gomega.HaveOccurred())
-			fakeOvn.asf.EventuallyExpectEmptyAddressSetExist(hostNetworkNamespace)
+
+			// check the namespace and ensure the address set being created with the right set of IPs in it.
+			// Note that NodeMgmtPortIP may or may not in the hostNetworkNamespace's address set, depends on
+			// if node1's subnet annotation already exists in the informer cache or not when hostNetwork
+			// namespace is handled in WatchNamespaces.
+			allowIPs := make([]string, 0, len(gwLRPIPs)+1)
+			for _, lrpIP := range gwLRPIPs {
+				allowIPs = append(allowIPs, lrpIP.IP.String())
+			}
+			fakeOvn.asf.ExpectAddressSetHasIPs(hostNetworkNamespace, allowIPs)
```
let me remove the check for now.

cathy-zhou commented:
@npinaeva @jcaamano @maiqueb please take a look of the updated code. @jcaamano I think I addressed all your comments except a few (I replied with my reasoning), and the one that we need to handle interaction between pod's `k8s.v1.cni.cncf.io/policy-for` annotation and NADs update. (we agreed to defer it to a later PR)

jcaamano commented on go-controller/pkg/ovn/gress_policy.go:
```diff
@@ -280,6 +290,7 @@ func (gp *gressPolicy) isEmpty() bool {
 // by the parent NetworkPolicy)
 // buildLocalPodACLs is safe for concurrent use, since it only uses gressPolicy fields that don't change
 // since creation, or are safe for concurrent use like peerVXAddressSets
+// Note that portGroupName is portGroup's name without network scope prefix
```
Not addressed?

cathy-zhou replied:
sorry, missed it. will fix.

jcaamano commented:
> @npinaeva @jcaamano @maiqueb please take a look of the updated code. @jcaamano I think I addressed all your comments except a few (I replied with my reasoning), and the one that we need to handle interaction between pod's `k8s.v1.cni.cncf.io/policy-for` annotation and NADs update. (we agreed to defer it to a later PR)

Answered some of your comments. Also this is still not addressed: https://github.com/ovn-org/ovn-kubernetes/pull/3382#discussion_r1164403160

Maybe worth waiting for @maiqueb to report a pass on his local tests?

maiqueb commented:
> > @npinaeva @jcaamano @maiqueb please take a look of the updated code. @jcaamano I think I addressed all your comments except a few (I replied with my reasoning), and the one that we need to handle interaction between pod's `k8s.v1.cni.cncf.io/policy-for` annotation and NADs update. (we agreed to defer it to a later PR)
> 
> Answered some of your comments. Also this is still not addressed: [#3382 (comment)](https://github.com/ovn-org/ovn-kubernetes/pull/3382#discussion_r1164403160)
> 
> Maybe worth waiting for @maiqueb to report a pass on his local tests?

It is now working for a pod selector based policy. I've pushed PR #3531 with it.
Adding more policies right now.

cathy-zhou commented:
/retest

npinaeva commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -58,6 +58,10 @@ type defaultDenyPortGroups struct {
 	policies map[string]bool
 }
 
+func (bnc *BaseNetworkController) getPortGroupName(readableGroupName string) string {
```
I don't if my comment is getting lost, I will comment here again
The point was to do both hashing and prefixing for port group name in one function like (the order of hashing and prefixing may be different)

func (ncni *NetworkControllerNetInfo) getPortGroupName(name string) string {
	networkScopedName := ncni.GetNetworkScopedName(name)
	return hashedPortGroup(networkScopedName)
}

that is to remove the logic of hashing/prefixing form the callers. 
We don't hash ClusterPortGroupName and ClusterRtrPortGroupName, and I think this function is a good place to unify this strange logic too, so it can skip hashing for these port groups in one place

For the same reason I suggest to pass only name to buildPortGroup, and then buildPortGroup itself may get the name of the port group by calling getPortGroupName
Please let me know if that makes sense

cathy-zhou replied:
talked to @npinaeva offline. It is hard to unify like she suggested as the defaultDenyPortGroup/policyPortGroup/multicastPortGoup all use different form. agreed to defer this until we agreed on a solution.

npinaeva replied:
I guess at this point this function is not needed, we will have a separate function for every type of port group to get its name?

npinaeva commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -182,21 +157,21 @@ func newDefaultNetworkControllerCommon(cnci *CommonNetworkControllerInfo,
 			controllerName:              DefaultNetworkControllerName,
-			NetConfInfo:                 &util.DefaultNetConfInfo{},
 			NetInfo:                     &util.DefaultNetInfo{},
+			NetConfInfo:                 &util.DefaultNetConfInfo{},
```
revert?

npinaeva commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -182,21 +157,21 @@ func newDefaultNetworkControllerCommon(cnci *CommonNetworkControllerInfo,
+			podSelectorAddressSets:      syncmap.NewSyncMap[*PodSelectorAddressSet](),
 			stopChan:                    defaultStopChan,
 			wg:                          defaultWg,
 		},
-		externalGWCache:        make(map[ktypes.NamespacedName]*externalRouteInfo),
-		exGWCacheMutex:         sync.RWMutex{},
-		networkPolicies:        syncmap.NewSyncMap[*networkPolicy](),
-		sharedNetpolPortGroups: syncmap.NewSyncMap[*defaultDenyPortGroups](),
-		podSelectorAddressSets: syncmap.NewSyncMap[*PodSelectorAddressSet](),
+		externalGWCache: make(map[ktypes.NamespacedName]*externalRouteInfo),
+		exGWCacheMutex:  sync.RWMutex{},
```
revert?

cathy-zhou replied:
didn't change it. it is only indentation change because we moved a few fields (sharedNetpolPortGroups etc.) into  BaseNetworkController.

npinaeva commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -1081,6 +991,13 @@ func (h *defaultNetworkControllerEventHandler) DeleteResource(obj, cachedObj int
 		ns := obj.(*kapi.Namespace)
 		return h.oc.deleteNamespace(ns)
 
+	case factory.PolicyType:
+		knp, ok := obj.(*knet.NetworkPolicy)
+		if !ok {
+			return fmt.Errorf("could not cast obj of type %T to *knet.NetworkPolicy", obj)
+		}
+		return h.oc.deleteNetworkPolicy(knp)
+
```
revert?

npinaeva commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -1129,6 +1037,9 @@ func (h *defaultNetworkControllerEventHandler) SyncFunc(objs []interface{}) erro
 		case factory.NamespaceType:
 			syncFunc = h.oc.syncNamespaces
 
+		case factory.PolicyType:
+			syncFunc = h.oc.syncNetworkPolicies
+
```
revert?

npinaeva commented on go-controller/pkg/ovn/gress_policy.go:
```diff
@@ -79,8 +85,10 @@ func (pp *portPolicy) getL4Match() (string, error) {
+func newGressPolicy(policyType knet.PolicyType, idx int, namespace, name, controllerName string, isNetPolStateless bool, netInfo util.NetInfo, netConfInfo util.NetConfInfo) *gressPolicy {
+	ipv4Mode, ipv6Mode := netConfInfo.IPMode()
 	return &gressPolicy{
+		NetInfo:           netInfo,
```
NetInfo is not used anymore?

npinaeva commented on go-controller/pkg/ovn/namespace_test.go:
```diff
@@ -295,7 +295,6 @@ var _ = ginkgo.Describe("OVN Namespace Operations", func() {
 
 			err = fakeOvn.controller.WatchNamespaces()
 			gomega.Expect(err).NotTo(gomega.HaveOccurred())
-			fakeOvn.asf.EventuallyExpectEmptyAddressSetExist(hostNetworkNamespace)
```
can you please comment on what changed here

npinaeva commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -494,3 +429,73 @@ func newEgressServiceController(client clientset.Interface, nbClient libovsdbcli
 		svcFactory.Core().V1().Nodes())
 }
+
+func (oc *DefaultNetworkController) syncNetworkPolicies(networkPolicies []interface{}) error {
```
I am not sure what is our common strategy, but I think it may be better to leave these netpol-specific function under base_network_controller_policy, otherwise all the default controller specific functions will flood this file, and then it is difficult to find them
wdyt @cathy-zhou @jcaamano ?

cathy-zhou replied:
probably to default_network_controller.go? this is not for base_network_controller, so it should not belong to base_network_controller_policy.go

npinaeva replied:
if we had proper packaging, every file would be split into bnc/<feature_name>, default_network_controller/<feature_name> and where needed same for the secondary controllers. Now when we put the default network controller parts to the same file (ovn or default_network_controller) it may difficult to split them back when/if we change the packages structure.
I agree that it doesn't make a lot of sense to put these function under bnc_ files, so we can make default_network_controller_policy file for that purpose, wdyt?

cathy-zhou replied:
create a file only for one function? I am not sure.

npinaeva replied:
I guess it should be based on if we are planning to have the same separate file in the future under default network controller package, or if we want to leave all functions in one big file . For now e.g. I am not sure what is the difference between `master.go` `ovn.go` and `default_network_controller.go`, so what is the right place for default network controller netpol functions (that also includes `addHairpinAllowACL`)
@jcaamano any thoughts?

jcaamano replied:
We can make default_network_controller_policy with WatchNetwoerkPolicies, addHairpinAllowACL and syncNetworkPolicies.

npinaeva reviewed (commented):
another round :)

jcaamano commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -519,74 +513,135 @@ func getPolicyType(policy *knet.NetworkPolicy) (bool, bool) {
-			errObjs = append(errObjs, pod)
+		// secondary networks
+		on, networkMap, err := util.GetPodNADToNetworkMapping(pod, bnc.NetInfo)
+		if err != nil || !on {
```
If you have the chance please add comment
```
// skip pods that are not on this network
if err != nil || !on {
```

and also my suggestion had a typo, sorry

s/remain/remaining

reminder there is two places for this

jcaamano commented on go-controller/pkg/ovn/multipolicy_test.go:
```diff
@@ -0,0 +1,487 @@
+
+func getExpectedDataPodsAndSwitchesForSecondaryNetwork(fakeOvn *FakeOVN, pods []testPod) []libovsdb.TestData {
+	data := []libovsdb.TestData{}
+	for _, ocInfo := range fakeOvn.secondaryControllers {
```
I am getting several warnings around this file that we are copying BaseNetworkController namespacesMutex which is not allowed and might not work as expected. I guess you should use pointers to the controllers.

cathy-zhou commented on go-controller/pkg/ovn/base_network_controller_multicast.go:
```diff
@@ -116,60 +117,63 @@ func getMulticastPortGroupName(namespace string) string {
 			continue
 		}
-		if portInfo, err := oc.logicalPortCache.get(pod, types.DefaultNetworkName); err != nil {
+		portInfoMap, err := bnc.logicalPortCache.getAll(pod)
```
it should for all NADs belong to this network controller

cathy-zhou commented:
@npinaeva @jcaamano again, please take a look.

npinaeva commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -519,74 +505,104 @@ func getPolicyType(policy *knet.NetworkPolicy) (bool, bool) {
-			continue
+		nadNames := bnc.getPodNADNames(pod)
+		for _, nadName := range nadNames {
+			portInfo, err := bnc.getNewLocalPolicyPortsForNAD(np, pod, nadName)
```
I guess at this point it makes sense to inline `getNewLocalPolicyPortsForNAD` back to this function, since it is not called from anywhere else?

npinaeva commented on go-controller/pkg/ovn/base_network_controller_multicast.go:
```diff
@@ -186,37 +190,37 @@ func deleteMulticastAllowPolicy(nbClient libovsdbclient.Client, ns string) error
 	// Remove old multicastDefaultDeny port group now that all ports
 	// have been added to the clusterPortGroup by WatchPods()
-	ops, err = libovsdbops.DeletePortGroupsOps(oc.nbClient, ops, legacyMulticastDefaultDenyPortGroup)
+	ops, err = libovsdbops.DeletePortGroupsOps(bnc.nbClient, ops, bnc.getPortGroupName(legacyMulticastDefaultDenyPortGroup))
```
we don't need to prefix stale port group right? since it was deleted before the networkPrefix is defined

cathy-zhou replied:
in that case, it is no-op. Otherwise, secondary network controllers would be trying to delete the legacy port groups of default network, which I don't want to happen. Another way is to do it only when `if !bnc.IsSecondary()` , but seems something we want to avoid as much as possible?

I personally is fine with either way.

cathy-zhou commented:
@npinaeva I made the change, please take a look.

npinaeva commented on go-controller/pkg/ovn/base_network_controller_multicast.go:
```diff
@@ -30,10 +29,17 @@ const (
 
-func getACLMatchAF(ipv4Match, ipv6Match string) string {
-	if config.IPv4Mode && config.IPv6Mode {
+func (bnc *BaseNetworkController) GetMulticastPortGroupName(base string) string {
```
I guess that should be getClusterPortGroupName? Because cluster port groups are not only used by multicast, but e.g. by egress firewall
 If so, I would suggest to move it to base_network_controller, update the name and add a comment for which base names it should be used

npinaeva commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -519,106 +501,124 @@ func getPolicyType(policy *knet.NetworkPolicy) (bool, bool) {
-		}
+		nadNames := bnc.getPodNADNames(pod)
+		for _, nadName := range nadNames {
+			var logicalPortName string
```
`logicalPortName := bnc.GetLogicalPortName(pod, nadName)` ?

cathy-zhou replied:
are you suggesting the change the function signature of GetSecondaryNetworkLogicalPortName(), or you meant to have one single function for both default and secondary network?

for latter, the problem is that we'd still need to check if it is default network, to decide weather or not prefix the logical port name. I believe that is the reason why we agreed to have two different functions.

npinaeva replied:
I mean this function already exists
```
func (bnc *BaseNetworkController) GetLogicalPortName(pod *kapi.Pod, nadName string) string {
	if !bnc.IsSecondary() {
		return util.GetLogicalPortName(pod.Namespace, pod.Name)
	} else {
		return util.GetSecondaryNetworkLogicalPortName(pod.Namespace, pod.Name, nadName)
	}
}
```
and seems to do exactly the same thing, so I just suggest using it

cathy-zhou replied:
Ah, I see. I didn't even remember we have this function. thanks!

npinaeva commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -519,106 +501,124 @@ func getPolicyType(policy *knet.NetworkPolicy) (bool, bool) {
-// if there are problems with fetching port info from logicalPortCache, pod will be added to errObjs.
-func (oc *DefaultNetworkController) getExistingLocalPolicyPorts(np *networkPolicy,
-	objs ...interface{}) (policyPortsToUUIDs map[string]string, policyPortUUIDs []string, errObjs []interface{}) {
+// if there are problems with fetching port info from logicalPortCache, error will be added to return error array.
```
looks like errors are not returned anymore, probably we can delete that comment line

npinaeva commented on go-controller/pkg/ovn/base_network_controller_policy.go:
```diff
@@ -519,106 +501,124 @@ func getPolicyType(policy *knet.NetworkPolicy) (bool, bool) {
-		portUUID := loadedPortUUID.(string)
+		nadNames := bnc.getPodNADNames(pod)
+		for _, nadName := range nadNames {
+			var logicalPortName string
+			if !bnc.IsSecondary() {
+				logicalPortName = util.GetLogicalPortName(pod.Namespace, pod.Name)
+			} else {
+				logicalPortName = util.GetSecondaryNetworkLogicalPortName(pod.Namespace, pod.Name, nadName)
+			}
```
`logicalPortName := bnc.GetLogicalPortName(pod, nadName)` ?

cathy-zhou replied:
same as above

npinaeva commented on go-controller/pkg/ovn/multicast_test.go:
```diff
@@ -52,7 +52,7 @@ func setIpMode(m ipMode) {
 }
 
-func getMulticastDefaultExpectedData(clusterPortGroup, clusterRtrPortGroup *nbdb.PortGroup) []libovsdb.TestData {
+func getMulticastDefaultExpectedData(fakeController *DefaultNetworkController, clusterPortGroup, clusterRtrPortGroup *nbdb.PortGroup) []libovsdb.TestData {
```
fakeController unused?

npinaeva commented on go-controller/pkg/ovn/multicast_test.go:
```diff
@@ -135,25 +135,25 @@ func getMulticastDefaultExpectedData(clusterPortGroup, clusterRtrPortGroup *nbdb
 
-func getMulticastDefaultStaleData(clusterPortGroup, clusterRtrPortGroup *nbdb.PortGroup) []libovsdb.TestData {
-	testData := getMulticastDefaultExpectedData(clusterPortGroup, clusterRtrPortGroup)
+func getMulticastDefaultStaleData(fakeController *DefaultNetworkController, clusterPortGroup, clusterRtrPortGroup *nbdb.PortGroup) []libovsdb.TestData {
```
fakeController will be unused, if `getMulticastDefaultExpectedData` is updated?

npinaeva commented on go-controller/pkg/ovn/multicast_test.go:
```diff
@@ -169,28 +169,28 @@ func getMulticastDefaultStaleData(clusterPortGroup, clusterRtrPortGroup *nbdb.Po
-	pg_hash := getMulticastPortGroupName(ns)
-	egressMatch := getACLMatch(pg_hash, getMulticastACLEgrMatch(), aclEgress)
+func getMulticastPolicyExpectedData(fakeController *DefaultNetworkController, ns string, ports []string) []libovsdb.TestData {
+	pg_hash := fakeController.getMulticastPortGroupName(ns)
```
we usually use
`fakeController = getFakeController(DefaultNetworkControllerName)` right here in the function to avoid adding arguments (of course in case the controller type is predefined)

npinaeva reviewed (commented):
Great progress, we are getting there :) 
Added some new comments, and here are some comments I don't see addressed:
https://github.com/ovn-org/ovn-kubernetes/pull/3382#discussion_r1168391464
https://github.com/ovn-org/ovn-kubernetes/pull/3382#discussion_r1166707318

cathy-zhou commented on go-controller/pkg/ovn/namespace.go:
```diff
@@ -396,74 +249,29 @@ func (oc *DefaultNetworkController) deleteNamespace(ns *kapi.Namespace) error {
+	if err != nil {
+		return nil, nil, err
 	}
+	defer oc.namespacesMutex.Unlock()
```
how, note taht in the two functions for default network controllers, configureNamespace() is receiver function of DefaultNetworkController. We need to somehow pass that to  ensureNamespaceLockedCommon(), which is strange to me.

npinaeva replied:
I am not sure I understand, being a method doesn't change function signature, so you still can do
```
func (bnc *BaseNetworkController) ensureNamespaceLockedCommon(ns string, readOnly bool, namespace *kapi.Namespace,
	ips []net.IP, configureNamespace func(nsInfo *namespaceInfo, ns *kapi.Namespace) error) (*kapi.Namespace, *namespaceInfo, func(), error) {...}

namespace, nsInfo, unlockFunc, err := oc.ensureNamespaceLockedCommon(ns, readOnly, namespace, ips, oc.configureNamespace)

namespace, nsInfo, unlockFunc, err := bsnc.ensureNamespaceLockedCommon(ns, readOnly, namespace, ips, bsnc.configureNamespaceCommon)
```

cathy-zhou replied:
you are right. fixed. Please take a look

npinaeva commented on go-controller/pkg/ovn/namespace_test.go:
```diff
@@ -67,8 +67,8 @@ func newNamespace(namespace string) *v1.Namespace {
 
-func getNsAddrSetHashNames(ns string) (string, string) {
-	return addressset.GetHashNamesForAS(getNamespaceAddrSetDbIDs(ns, DefaultNetworkControllerName))
+func getNsAddrSetHashNames(ns, controllerName string) (string, string) {
```
I would say let's leave this function as is, and make a new one like `getMultinetNsAddrSetHashNames(ns, controllerName string)`, and it will have only 1 usage from `getGressACLs`, and will reduce the number of required changes

npinaeva commented on go-controller/pkg/ovn/multipolicy_test.go:
```diff
@@ -0,0 +1,487 @@
+				startOvn(initialDB, []v1.Namespace{namespace1, namespace2}, nil, nil,
+					[]nettypes.NetworkAttachmentDefinition{*nad}, nil, nil)
+
+				ocInfo, ok := fakeOvn.secondaryControllers["network1"]
+				gomega.Expect(ok).To(gomega.Equal(true))
+				asf := ocInfo.asf
+				gomega.Expect(asf).NotTo(gomega.Equal(nil))
+				gomega.Expect(asf.ControllerName).To(gomega.Equal("network1-network-controller"))
```
maybe we can make it a part of `startOvn`, so it checks expected secondary controllers are created based on given nads?

npinaeva commented on go-controller/pkg/ovn/policy_test.go:
```diff
@@ -80,9 +69,17 @@ func newNetworkPolicy(name, namespace string, podSelector metav1.LabelSelector,
+	}
+}
+
 func getDefaultDenyData(networkPolicy *knet.NetworkPolicy, ports []string,
```
If we merge https://github.com/ovn-org/ovn-kubernetes/pull/3540 soon, we will be able to change the signature of this function, then pass default network args for existing wrappers, and create new functions like `getDefaultDenyDataMultinet` and `getPolicyDataMultinet`.
In that case calls in this file won't need to be changed

npinaeva commented on go-controller/pkg/ovn/base_network_controller_multicast.go:
```diff
@@ -186,37 +190,37 @@ func deleteMulticastAllowPolicy(nbClient libovsdbclient.Client, ns string) error
 	// Remove old multicastDefaultDeny port group now that all ports
 	// have been added to the clusterPortGroup by WatchPods()
-	ops, err = libovsdbops.DeletePortGroupsOps(oc.nbClient, ops, legacyMulticastDefaultDenyPortGroup)
+	ops, err = libovsdbops.DeletePortGroupsOps(bnc.nbClient, ops, bnc.getClusterPortGroupName(legacyMulticastDefaultDenyPortGroup))
```
I think maybe in that case it makes sense to do
```
if !bnc.IsSecondary() {
		ops, err = libovsdbops.DeletePortGroupsOps(bnc.nbClient, ops, legacyMulticastDefaultDenyPortGroup)
		if err != nil {
			return err
		}
	}
```
to avoid confusion with prefixing

npinaeva reviewed (commented):
All the previous comments look addressed, just a couple more for the unit test commit.
It's great to have a unit test with both multi and regular netpols, thanks for implementing that @cathy-zhou !

cathy-zhou commented:
/retest

npinaeva commented on go-controller/pkg/ovn/egressfirewall_test.go:
```diff
@@ -130,7 +130,7 @@ var _ = ginkgo.Describe("OVN EgressFirewall Operations", func() {
 				app.Action = func(ctx *cli.Context) error {
 					// owned by non-existing namespace
-					fakeController := getFakeController(DefaultNetworkControllerName)
+					fakeController := &DefaultNetworkController{BaseNetworkController: *getFakeBaseController(&util.DefaultNetInfo{}, &util.DefaultNetConfInfo{})}
```
rollback?

npinaeva reviewed (commented):
one last nit, otherwise
/lgtm
thanks @cathy-zhou!
